src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -20,21 +22,28 @@ internal sealed class FileChangeWatcher
 
         /// <summary>
         /// We create a queue of tasks against the IVsFileChangeEx service for two reasons. First, we are obtaining the service asynchronously, and don't want to
-        /// block on it being available, so anybody who wants to do anything must wait for it. Secondly, the service itself is single-threaded; in the past
-        /// we've blocked up a bunch of threads all trying to use it at once. If the latter ever changes, we probably want to reconsider the implementation of this.
+        /// block on it being available, so anybody who wants to do anything must wait for it. Secondly, the service itself is single-threaded; the entry points
+        /// are asynchronous so we avoid starving the thread pool, but there's still no reason to create a lot more work blocked than needed. Finally, since this
+        /// is all happening async, we generally need to ensure that an operation that happens for an earlier call to this is done before we do later calls. For example,
+        /// if we started a subscription for a file, we need to make sure that's done before we try to unsubscribe from it.
         /// For performance and correctness reasons, NOTHING should ever do a block on this; figure out how to do your work without a block and add any work to
         /// the end of the queue.
         /// </summary>
-        private Task<IVsFileChangeEx> _taskQueue;
-        private static readonly Func<Task<IVsFileChangeEx>, object, IVsFileChangeEx> _executeActionDelegate =
-            (precedingTask, state) => { ((Action<IVsFileChangeEx>)state)(precedingTask.Result); return precedingTask.Result; };
+        private Task<IVsAsyncFileChangeEx> _taskQueue;
+        private static readonly Func<Task<IVsAsyncFileChangeEx>, object, Task<IVsAsyncFileChangeEx>> _executeActionDelegate =
+            async (precedingTask, state) =>
+            {
+                var action = (Func<IVsAsyncFileChangeEx, Task>)state;
+                await action(precedingTask.Result).ConfigureAwait(false);
```

Comment:
nit.  i feel like this should be wrapped in some sort of try/catch.  my brain is borking on trying to figure otu the implications of this throwing and what does for this queue.

Reply:
Uh, if this is throwing...we probably should just crash. The assumption is the underlying service is resilient -- and this assumption isn't really new.